File: frontend/olympic_it_project/olympic_it_project/lib/core/api_client.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:http_parser/http_parser.dart';
import 'package:olympic_it_project/core/config.dart'; 
import 'package:olympic_it_project/core/storage_token.dart';
import 'package:olympic_it_project/service/auth_service.dart';

class ApiClient {
  static final ApiClient instance = ApiClient._internal();
  ApiClient._internal();

  static const String host = HOST;
  final String _baseUrl = API_BASE;
  
  Future<void>? _refreshFuture;

  Future<void> _refreshOnce() {
    _refreshFuture ??= AuthService().refreshTokens().whenComplete(() {
      _refreshFuture = null;
    });
    return _refreshFuture!;
  }

  bool _shouldRefresh(String endpoint) {
    const excludedEndpoints = {
      "auth/login",
      "auth/resend-otp",
      "auth/forgot-password",
      "auth/reset-password",
      "auth/refresh",
    };
    return !excludedEndpoints.contains(endpoint);
  }

  Future<Map<String, String>> _getHeaders() async {
    final token = await StorageToken.instance.getAccessToken();
    return {
      'Content-Type': 'application/json',
      if (token != null) 'Authorization': 'Bearer $token',
    };
  }

  Future<http.Response> postRaw(String endpoint, dynamic body) async =>
      await http
          .post(
            Uri.parse(_baseUrl + endpoint),
            headers: await _getHeaders(),
            body: jsonEncode(body),
          )
          .timeout(const Duration(seconds: 15));

  Future<http.Response> post(String endpoint, dynamic body) async {
    var response = await postRaw(endpoint, body);
    if (response.statusCode == 401 && _shouldRefresh(endpoint)) {
      await _refreshOnce();
      response = await postRaw(endpoint, body);
    }
    return response;
  }

  Future<http.Response> getRaw(String endpoint) async => await http
      .get(Uri.parse(_baseUrl + endpoint), headers: await _getHeaders())
      .timeout(const Duration(seconds: 15));

  Future<http.Response> get(String endpoint) async {
    var response = await getRaw(endpoint);
    if (response.statusCode == 401 && _shouldRefresh(endpoint)) {
      await _refreshOnce();
      response = await getRaw(endpoint);
    }
    return response;
  }

  Future<http.Response> putRaw(String endpoint, dynamic body) async =>
      await http
          .put(
            Uri.parse(_baseUrl + endpoint),
            headers: await _getHeaders(),
            body: jsonEncode(body),
          )
          .timeout(const Duration(seconds: 15));

  Future<http.Response> put(String endpoint, dynamic body) async {
    var response = await putRaw(endpoint, body);
    if (response.statusCode == 401 && _shouldRefresh(endpoint)) {
      await _refreshOnce();
      response = await putRaw(endpoint, body);
    }
    return response;
  }

  Future<http.Response> deleteRaw(String endpoint) async => await http
      .delete(Uri.parse(_baseUrl + endpoint), headers: await _getHeaders())
      .timeout(const Duration(seconds: 15));

  Future<http.Response> delete(String endpoint) async {
    var response = await deleteRaw(endpoint);
    if (response.statusCode == 401 && _shouldRefresh(endpoint)) {
      await _refreshOnce();
      response = await deleteRaw(endpoint);
    }
    return response;
  }

  // ================= DELETE có body (cần cho các API xóa kèm dữ liệu) =================
  Future<http.Response> deleteWithBodyRaw(
    String endpoint,
    dynamic body,
  ) async => await http
      .delete(
        Uri.parse(_baseUrl + endpoint),
        headers: await _getHeaders(),
        body: jsonEncode(body),
      )
      .timeout(const Duration(seconds: 15));

  Future<http.Response> deleteWithBody(String endpoint, dynamic body) async {
    var response = await deleteWithBodyRaw(endpoint, body);
    if (response.statusCode == 401 && _shouldRefresh(endpoint)) {
      await _refreshOnce();
      response = await deleteWithBodyRaw(endpoint, body);
    }
    return response;
  }

  Future<http.StreamedResponse> uploadMultipartRaw(
    String endpoint,
    String field,
    String filePath,
  ) async {
    final token = await StorageToken.instance.getAccessToken();
    final request = http.MultipartRequest(
      "POST",
      Uri.parse(_baseUrl + endpoint),
    );

    if (token != null) {
      request.headers["Authorization"] = "Bearer $token";
    }

    final file = File(filePath);
    final fileName = file.path.split('/').last;
    final ext = fileName.split('.').last.toLowerCase();

    final contentType = (ext == "png" || ext == "jpg" || ext == "jpeg")
        ? "image/$ext"
        : "video/mp4";

    request.files.add(
      await http.MultipartFile.fromPath(
        field,
        filePath,
        filename: fileName,
        contentType: MediaType.parse(contentType),
      ),
    );

    return request.send();
  }

  Future<http.StreamedResponse> uploadMultipart(
    String endpoint,
    String field,
    String filePath,
  ) async {
    var response = await uploadMultipartRaw(endpoint, field, filePath);

    if (response.statusCode == 401 && _shouldRefresh(endpoint)) {
      await _refreshOnce();
      response = await uploadMultipartRaw(endpoint, field, filePath);
    }
    return response;
  }
}

dynamic safeDecode(http.Response response) {
  if (response.body.trim().isEmpty) {
    return {
      "code": response.statusCode,
      "message": "Empty response body",
      "data": null,
    };
  }
  try {
    return jsonDecode(response.body);
  } catch (e) {
    return {
      "code": response.statusCode,
      "message": "Không thể phân tích phản hồi từ server",
      "data": null,
    };
  }
}

File: admin-web/js/core/api_client.js
import { API_BASE } from "./config.js";
import { storageToken } from "./storage_token.js";

const EXCLUDED_FROM_REFRESH = new Set([
    "auth/login",
    "auth/resend-otp",
    "auth/forgot-password",
    "auth/reset-password",
    "auth/refresh",
]);

class ApiClient {
    _authService = null;

    init(authService) {
        this._authService = authService;
    }

    _buildUrl(endpoint) {
        const path = endpoint.startsWith("/") ? endpoint.slice(1) : endpoint;
        return new URL(path, API_BASE).toString();
    }

    _headers(endpoint) {
        const token = storageToken.getAccessToken();
        const skipAuth = EXCLUDED_FROM_REFRESH.has(endpoint);
        return {
            "Content-Type": "application/json",
            ...(token && !skipAuth ? { Authorization: `Bearer ${token}` } : {}),
        };
    }

    _shouldRefresh(endpoint) {
        return !EXCLUDED_FROM_REFRESH.has(endpoint);
    }

    async _requestRaw(endpoint, method, body) {
        return fetch(this._buildUrl(endpoint), {
            method,
            headers: this._headers(endpoint),
            body: body !== undefined ? JSON.stringify(body) : undefined,
        });
    }

    async _requestWithRetry(endpoint, method, body) {
        let response = await this._requestRaw(endpoint, method, body);

        if (response.status === 401 && this._shouldRefresh(endpoint) && this._authService) {
            await this._authService.refreshTokens();
            response = await this._requestRaw(endpoint, method, body);
        }

        return response;
    }

    get(endpoint) {
        return this._requestWithRetry(endpoint, "GET");
    }

    post(endpoint, body) {
        return this._requestWithRetry(endpoint, "POST", body ?? {});
    }

    put(endpoint, body) {
        return this._requestWithRetry(endpoint, "PUT", body ?? {});
    }

    delete(endpoint) {
        return this._requestWithRetry(endpoint, "DELETE");
    }

    deleteWithBody(endpoint, body) {
        return this._requestWithRetry(endpoint, "DELETE", body ?? {});
    }

    postRaw(endpoint, body) {
        return this._requestRaw(endpoint, "POST", body ?? {});
    }

    async uploadMultipart(endpoint, field, file) {
        const token = storageToken.getAccessToken();
        const form = new FormData();
        form.append(field, file, file.name);

        return fetch(this._buildUrl(endpoint), {
            method: "POST",
            headers: token ? { Authorization: `Bearer ${token}` } : {},
            body: form,
        });
    }
}

export const apiClient = new ApiClient();

export async function safeDecode(response) {
    const text = (await response.text()).trim();

    if (!response.ok) {
        throw new Error(`Lỗi API: ${response.status} - ${text}`);
    }

    if (!text) {
        return {
            code: 200,
            message: "Success",
            data: null,
        };
    }

    return JSON.parse(text);
}